File: frontend/src/api/hierarchy.ts
import { apiClient } from './client'
import type { HierarchyData, CategoryEntry, SubCategoryEntry, ItemEntry } from '../types'

export async function fetchHierarchy(): Promise<HierarchyData> {
  const response = await apiClient.get<HierarchyData>('/hierarchy/all')
  return response.data
}

export async function fetchCategories(): Promise<CategoryEntry[]> {
  const response = await apiClient.get<CategoryEntry[]>('/hierarchy/categories')
  return response.data
}

export async function fetchSubCategories(): Promise<SubCategoryEntry[]> {
  const response = await apiClient.get<SubCategoryEntry[]>('/hierarchy/subcategories')
  return response.data
}

export async function fetchItems(): Promise<ItemEntry[]> {
  const response = await apiClient.get<ItemEntry[]>('/hierarchy/items')
  return response.data
}

export async function updateCategories(entries: CategoryEntry[]): Promise<void> {
  await apiClient.put('/hierarchy/categories', { entries })
}

export async function updateSubCategories(entries: SubCategoryEntry[]): Promise<void> {
  await apiClient.put('/hierarchy/subcategories', { entries })
}

export async function updateItems(entries: ItemEntry[]): Promise<void> {
  await apiClient.put('/hierarchy/items', { entries })
}

File: backend/schemas/hierarchy.py
"""Pydantic schemas for hierarchy data."""
from typing import List, Optional

from pydantic import BaseModel, Field


class CategoryEntry(BaseModel):
    """Top-level category."""
    sub_category: str = Field(..., alias="Sub Category")
    category: str = Field(..., alias="Category")

    class Config:
        populate_by_name = True


class SubCategoryEntry(BaseModel):
    """Middle-level sub-category."""
    sub_sub_category: str = Field(..., alias="Sub Sub Category")
    sub_category: str = Field(..., alias="Sub Category")

    class Config:
        populate_by_name = True


class ItemEntry(BaseModel):
    """Base-level item."""
    all_items: str = Field(..., alias="All Items")
    sub_sub_category: str = Field(..., alias="Sub Sub Category")

    class Config:
        populate_by_name = True


class HierarchyUpdate(BaseModel):
    """Batch update for hierarchy tables."""
    entries: List[dict]


class HierarchyData(BaseModel):
    """Complete hierarchy data."""
    categories: List[CategoryEntry]
    sub_categories: List[SubCategoryEntry]
    items: List[ItemEntry]
